tests/integration_tests/test_server.py

```diff
@@ -0,0 +1,164 @@
+from tests.integration_tests.utils import text_feature
+from tests.integration_tests.utils import categorical_feature
+from tests.integration_tests.utils import numerical_feature
+from tests.integration_tests.utils import generate_data
+from tests.integration_tests.utils import sequence_feature
+from tests.integration_tests.utils import random_string
+
+
+# The following imports are pytest fixtures, required for running the tests
+from tests.fixtures.filenames import csv_filename
+from ludwig.serve import server, ALL_FEATURES_PRESENT_ERROR
+from starlette.testclient import TestClient
+from starlette.datastructures import UploadFile
+import os
+
+
+def train_model(input_features, output_features, data_csv):
+    """
+    Helper method to avoid code repetition in running an experiment
+    :param input_features: input schema
+    :param output_features: output schema
+    :param data_csv: path to data
+    :return: None
+    """
+    model_definition = {
+        'input_features': input_features,
+        'output_features': output_features,
+        'combiner': {'type': 'concat', 'fc_size': 14},
+        'training': {'epochs': 2}
+    }
+
+    model = LudwigModel(model_definition)
+
+    # Training with csv
+    model.train(
+        data_csv=data_csv,
+        skip_save_processed_input=True,
+        skip_save_progress=True,
+        skip_save_unprocessed_output=True
+    )
+
+    model.predict(data_csv=data_csv)
+
+    # Remove results/intermediate data saved to disk
+    shutil.rmtree(model.exp_dir_name, ignore_errors=True)
+
+    # Training with dataframe
+    data_df = read_csv(data_csv)
+    model.train(
+        data_df=data_df,
+        skip_save_processed_input=True,
+        skip_save_progress=True,
+        skip_save_unprocessed_output=True
+    )
+    model.predict(data_df=data_df)
+    return model
+
+
+def example_input(input_features):
+    data = {}
+    for i in input_features:
+        name = i['name']
+        if i['type'] in ['image', 'video', 'audio']:
+            data[name] = random_string()
+        else:
+            data[name] = random_string()
+    return data
+
+
+def output_keys_for(output_features):
+    keys = []
+    for f in output_features:
+        name = f['name']
+        if f['type'] == 'category':
+            keys.append("{}_predictions".format(name))
+            keys.append("{}_probability".format(name))
+            keys.append("{}_probabilities_<UNK>".format(name))
+            for category in f['idx2str']:
+                keys.append(
+                    "{}_probabilities_{}".format(name, category))
+
+        elif f['type'] == 'numerical':
+            keys.append("{}_predictions".format(name))
+        else:
+            raise NotImplementedError
+    return keys
+
+
+def convert_to_form(entry):
+    data = {}
+    files = []
+    for k, v in entry.items():
+        if type(v) == str and os.path.exists(v):
+            files.append((k, (v, open(v, 'rb'), 'image/jpeg')))
```

Comment:
these files are open but never closed. They are actually removed inside the `/predict` but I'm not sure, probably we should be closing them too, ideally right before removing them in serve.py:74. What do you think?

Reply:
These files are consumed by the test client and get converted to `starlette.UploadFile`s by the time they get to serve.py. Will close inline